File: src/handlers/lib/reset/reset.rs
use crate::{config::Config, crypto::secure_token, handlers::Error, mailer::send_email, models::user::get_by_email_or_phone_number, sms::send_sms};
use chrono::Utc;
use diesel::{
    pg::PgConnection,
    r2d2::{ConnectionManager, PooledConnection},
    Connection,
};
use log::error;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize)]
pub struct ResetForm {
    pub username: String,
}

pub fn reset(config: &Config, connection: &PooledConnection<ConnectionManager<PgConnection>>, reset_form: ResetForm) -> Result<(), Error> {
    if !config.email_rule.is_match(&reset_form.username) && !config.phone_rule.is_match(&reset_form.username) {
        return Err(Error::new(
            422,
            json!({"code": "email_or_phone_number_required"}),
            "Email or Phone Number Is Required To Recover Account".to_string(),
        ));
    }

    let transaction = connection.transaction::<_, Error, _>(|| {
        let user = get_by_email_or_phone_number(reset_form.username.clone(), reset_form.username.clone(), connection);

        if user.is_err() {
            let err = user.err().unwrap();
            error!("{:?}", err);
            return Ok(());
        }

        let mut user = user.unwrap();

        user.recovery_token_sent_at = Some(Utc::now().naive_utc());

        if config.email_rule.is_match(&reset_form.username) && user.email.is_some() && user.email_confirmed {
            if !user.email_confirmed {
                return Ok(());
            }

            let template = config.clone().get_recovery_email_template();

            user.recovery_token = Some(secure_token(100));

            let user = user.save(&connection);

            if user.is_err() {
                let err = user.err().unwrap();
                error!("{:?}", err);
                return Err(Error::from(err));
            }

            let user = user.unwrap();

            let data = json!({
                "recovery_token": user.recovery_token.clone().unwrap(),
                "site_url": config.site_url,
                "email": user.email
            });

            let email = send_email(template, data, user.email.unwrap(), config);

            if email.is_err() {
                let err = email.err().unwrap();

                error!("{:?}", err);

                return Err(Error::from(err));
            }

            return Ok(());
        } else if config.phone_rule.is_match(&reset_form.username) && user.phone_number.is_some() && user.phone_confirmed {
            user.recovery_token = Some(secure_token(6));

            let user = user.save(&connection);

            if user.is_err() {
                let err = user.err().unwrap();
                error!("{:?}", err);
                return Err(Error::from(err));
            }

            let user = user.unwrap();

            let template = config.clone().get_recovery_sms_template();

            let data = json!({
                "recovery_token": user.recovery_token.clone().unwrap(),
                "site_url": config.site_url,
                "phone_number": user.phone_number
            });

            let sms = send_sms(template, data, user.phone_number.unwrap(), config);

            if sms.is_err() {
                let err = sms.err().unwrap();

                error!("{:?}", err);

                return Err(Error::from(err));
            }

            return Ok(());
        }

        return Ok(());
    });

    if transaction.is_err() {
        let err = transaction.err().unwrap();
        error!("{:?}", err);
        return Err(err);
    }

    return Ok(());
}

File: handlers/lib/reset/reset.go
package reset

import (
	"time"

	"github.com/sirupsen/logrus"
	"github.com/thanhpk/randstr"
	"github.com/zolamk/trust/config"
	"github.com/zolamk/trust/handlers"
	"github.com/zolamk/trust/lib/email"
	"github.com/zolamk/trust/lib/sms"
	"github.com/zolamk/trust/model"
	"gorm.io/gorm"
)

func Reset(db *gorm.DB, config *config.Config, username string) (bool, error) {

	now := time.Now()

	user := &model.User{}

	if tx := db.First(user, "phone = ? or email = ?", username, username); tx.Error != nil {

		if tx.Error == gorm.ErrRecordNotFound {
			return true, nil
		}

		logrus.Error(tx.Error)

		return false, handlers.ErrInternal

	}

	err := db.Transaction(func(tx *gorm.DB) error {

		if user.Email != nil && *user.Email == username && user.EmailConfirmed {

			if user.RecoveryTokenSentAt != nil && time.Since(*user.RecoveryTokenSentAt).Minutes() < float64(config.MinutesBetweenResend) {
				return nil
			}

			token := randstr.String(100)

			user.RecoveryToken = &token

			user.RecoveryTokenSentAt = &now

			if err := user.Save(db); err != nil {
				logrus.Error(err)
				return handlers.ErrInternal
			}

			context := &map[string]string{
				"site_url":             config.SiteURL,
				"email_recovery_token": *user.RecoveryToken,
				"instance_url":         config.InstanceURL,
			}

			if err := email.SendEmail(config.RecoveryTemplate, context, user.Email, config); err != nil {
				logrus.Error(err)
				return handlers.ErrInternal
			}

		}

		if user.Phone != nil && *user.Phone == username && user.PhoneConfirmed {

			if user.RecoveryTokenSentAt != nil && time.Since(*user.RecoveryTokenSentAt).Minutes() < float64(config.MinutesBetweenResend) {
				return nil
			}

			token := randstr.String(6)

			user.RecoveryToken = &token

			user.RecoveryTokenSentAt = &now

			if err := user.Save(db); err != nil {
				logrus.Error(err)
				return handlers.ErrInternal
			}

			context := &map[string]string{
				"site_url":             config.SiteURL,
				"phone_recovery_token": *user.RecoveryToken,
				"instance_url":         config.InstanceURL,
			}

			if err := sms.SendSMS(config.RecoveryTemplate, user.Phone, context, config.SMS); err != nil {
				logrus.Error(err)
				return handlers.ErrInternal
			}

		}

		return nil

	})

	return true, err

}
